### PR TITLE
Refactor dmChannelWatchInfo to queryCoord meta

### PR DESCRIPTION
**Pull request by xige-16**
fix issue: #12952 #12190 #12340
Signed-off-by: xige-16 <xi.ge@zilliz.com>

### CODE REVIEW

File: internal/querycoord/cluster.go [hunk 1]
```diff
@@ -308,15 +276,18 @@ func (c *queryNodeCluster) watchDmChannels(ctx context.Context, nodeID int64, in
 			log.Debug("watchDmChannels: queryNode watch dm channel error", zap.String("error", err.Error()))
 			return err
 		}
-		channels := make([]string, 0)
+		dmChannelWatchInfo := make([]*querypb.DmChannelWatchInfo, 0)
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/querycoord/cluster.go]: fixed

File: internal/querycoord/meta.go [hunk 2]
```diff
@@ -1177,55 +1138,71 @@ func saveGlobalCollectionInfo(collectionID UniqueID, info *querypb.CollectionInf
...
 		}
 
-		key := fmt.Sprintf("%s/%d/%s", deltaChannelMetaPrefix, collectionID, info.ChannelName)
+		key := fmt.Sprintf("%s/%d/%s", dmChannelMetaPrefix, info.CollectionID, info.DmChannel)
 		kvs[key] = string(infoBytes)
 	}
 	return kv.MultiSave(kvs)
 }
+
+func removeMetaByCollectionID(collectionID UniqueID, kv kv.MetaKv) error {
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/querycoord/meta.go]: fixed

File: internal/querycoord/meta.go [hunk 3]
```diff
@@ -1177,55 +1138,71 @@ func saveGlobalCollectionInfo(collectionID UniqueID, info *querypb.CollectionInf
...
+	prefixes = append(prefixes, segmentInfosPrefix)
+	dmChannelInfosPrefix := fmt.Sprintf("%s/%d", dmChannelMetaPrefix, collectionID)
+	prefixes = append(prefixes, dmChannelInfosPrefix)
+	deltaChannelInfosPrefix := fmt.Sprintf("%s/%d", deltaChannelMetaPrefix, collectionID)
+	prefixes = append(prefixes, deltaChannelInfosPrefix)
+
+	return kv.MultiRemoveWithPrefix(prefixes)
+}
+
+func removeMetaByPartitionID(collectionID UniqueID, saveCollectionInfo *querypb.CollectionInfo, removePartitionIDs []UniqueID, kv kv.MetaKv) error {
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/querycoord/meta.go]: fixed

File: internal/querycoord/channel_allocator.go [hunk 4]
```diff
@@ -31,13 +33,17 @@ func defaultChannelAllocatePolicy() ChannelAllocatePolicy {
...
-func shuffleChannelsToQueryNode(ctx context.Context, reqs []*querypb.WatchDmChannelsRequest, cluster Cluster, wait bool, excludeNodeIDs []int64) error {
+func shuffleChannelsToQueryNode(ctx context.Context, reqs []*querypb.WatchDmChannelsRequest, cluster Cluster, metaCache Meta, wait bool, excludeNodeIDs []int64) error {
+	if len(reqs) == 0 {
+		return nil
+	}
 	for {
-		availableNodes, err := cluster.onlineNodes()
-		if err != nil {
-			log.Debug(err.Error())
+		availableNodes := cluster.onlineNodes()
```
Comment (xiaofan-luan) [inline internal/querycoord/channel_allocator.go:43]: There is a inconsitency here. 
clusert.OfflineNodes return error if no node available.
cluster.onlineNodes return only empty list and the err is handled outside, which mode do you prefer?

File: internal/querycoord/channel_allocator.go [hunk 5]
```diff
@@ -31,13 +33,17 @@ func defaultChannelAllocatePolicy() ChannelAllocatePolicy {
...
+		if len(availableNodes) == 0 {
```
Comment (xiaofan-luan) [inline internal/querycoord/channel_allocator.go:44]: need another log, print out onlineNodes, excludeNodes and there channel infos.

File: internal/querycoord/cluster.go [hunk 6]
```diff
@@ -42,27 +41,22 @@ import (
 )
 
 const (
-	queryNodeMetaPrefix = "queryCoord-queryNodeMeta"
```
Comment (xiaofan-luan) [inline internal/querycoord/cluster.go:45]: remove queryNodeMetaPrefix is very clever, make the whole thing much clear.

File: internal/querycoord/task.go [hunk 7]
```diff
@@ -320,60 +318,29 @@ func (lct *loadCollectionTask) execute(ctx context.Context) error {
 	defer cancel2()
```
Comment (xiaofan-luan) [inline internal/querycoord/task.go:318]: functions loadCollectionTask execute is too long, better to seperate into multiple part

File: internal/querycoord/task.go [hunk 8]
```diff
@@ -320,60 +318,29 @@ func (lct *loadCollectionTask) execute(ctx context.Context) error {
...
 
 	loadSegmentReqs := make([]*querypb.LoadSegmentsRequest, 0)
 	watchDmChannelReqs := make([]*querypb.WatchDmChannelsRequest, 0)
-	channelsToWatch := make([]string, 0)
-	var watchDeltaChannels []*datapb.VchannelInfo
+	var deltaChannelInfos []*datapb.VchannelInfo
+	var dmChannelInfos []*datapb.VchannelInfo
 	for _, partitionID := range toLoadPartitionIDs {
-		dmChannelInfos, binlogs, err := getRecoveryInfo(lct.ctx, lct.dataCoord, collectionID, partitionID)
+		vChannelInfos, binlogs, err := getRecoveryInfo(lct.ctx, lct.dataCoord, collectionID, partitionID)
```
Comment (xiaofan-luan) [inline internal/querycoord/task.go:341]: Probably  need to batch getRecoveryInfo if partition id is a long list, so far I'm ok with this

File: internal/querycoord/segment_allocator.go [hunk 9]
```diff
@@ -80,21 +77,23 @@ func shuffleSegmentsToQueryNode(ctx context.Context, reqs []*querypb.LoadSegment
...
 				req.DstNodeID = nodeIDSlice[0]
-				nodeID2NumSegemnt[nodeIDSlice[0]]++
+				nodeID2NumSegment[nodeIDSlice[0]]++
 			}
 			return nil
 		}
 
 		if !wait {
-			return errors.New("no queryNode to allocate")
+			err := errors.New("no queryNode to allocate")
```
Comment (xiaofan-luan) [inline internal/querycoord/segment_allocator.go:89]: this error information lack of information.
why no querynode?
1) dead?
2) lack of memory?
3) or ..

Comment (xiaofan-luan) [reply internal/querycoord/segment_allocator.go:89]: it will help on locating the root cause

**Comment by xiaofan-luan** [comment]
overall is good~ the logic is clear, all collection related meta is stored in meta and node related is on  queryNodeInfoPrefix.

File: internal/querycoord/channel_allocator.go (hunk 4 above)
Comment (xige-16) [reply internal/querycoord/channel_allocator.go:43]: I prefer the latter, and it would be more flexible to handle the error outside.

**Comment by xige-16** [comment]
rerun failure checks

**Comment by xige-16** [comment]
rerun failure checks

File: internal/querycoord/channel_allocator.go (hunk 5 above)
Comment (xige-16) [reply internal/querycoord/channel_allocator.go]: fixed

File: internal/querycoord/segment_allocator.go (hunk 9 above)
Comment (xige-16) [reply internal/querycoord/segment_allocator.go]: more error info has been added

**Pull request merged**